fxrunner/src/lib/splash.rs

```diff
@@ -0,0 +1,252 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::ffi::CStr;
+use std::io;
+use std::ptr::null_mut;
+use std::thread;
+
+use async_trait::async_trait;
+use lazy_static::lazy_static;
+use tokio::sync::oneshot;
+use winapi::shared::minwindef::{DWORD, LPARAM, LRESULT, UINT, WPARAM};
+use winapi::shared::windef::HWND;
+use winapi::shared::winerror;
+use winapi::um::winuser::{MSG, WNDCLASSA};
+use winapi::um::{errhandlingapi, processthreadsapi, wingdi, winuser};
+
+use crate::osapi::error::{check_nonnull, check_nonzero};
+use crate::osapi::handle::{get_module_handle, ModuleHandle};
+
+lazy_static! {
+    static ref WINDOW_CLASS_NAME: &'static CStr =
+        CStr::from_bytes_with_nul(&b"fxrunnerbg\0"[..]).unwrap();
+}
+
+const MESSAGE_CLOSE_SPLASH: UINT = winuser::WM_USER + 1;
+
+#[async_trait]
+pub trait Splash: Sized {
+    async fn new(display_widht: u32, display_height: u32) -> Result<Self, io::Error>;
+    fn destroy(&mut self) -> Result<(), io::Error>;
+}
+
+/// A splash screen that covers the entire display.
+///
+/// The splash screen is painted a solid red (#FF0000) so that the Firefox Window
+/// can be easily differentiated from the background.
+pub struct WindowsSplash {
+    /// The thread ID of the UI thread, so that we may send messages to it via
+    /// `PostThreadMessageA` API.
+    ui_thread_id: DWORD,
+
+    /// The join handle for the thread.
+    ui_thread_join_handle: Option<thread::JoinHandle<()>>,
+}
+
+//
+#[async_trait]
+impl Splash for WindowsSplash {
+    /// Create a new `Splash` with the given width and height.
+    async fn new(display_width: u32, display_height: u32) -> Result<WindowsSplash, io::Error> {
+        let (tx, rx) = oneshot::channel::<Result<DWORD, io::Error>>();
+
+        let join_handle = thread::spawn(move || {
+            let window_handle = match create_and_show_window(display_width, display_height) {
```

Comment:
Nit: may be worth noting in a comment that we have to do this dance of creating the window on a background thread and sending success or failure over a channel because the event loop for a given window has to be checked on the thread that created that window.